### PR TITLE
Adds initial uncertainty analysis capability by refactoring

### PR DESCRIPTION
**Pull request by rjhanes**
Need to merge carefully here

The data branch that works with this refactoring (including updated config files) is tiny-uncertaintypoc

You'll likely need to create a directory called 'results' in your data folder prior to running.

### CODE REVIEW

File: celavi/scenario.py [hunk 1]
```diff
@@ -0,0 +1,569 @@
...
+
+    def __init__(self, parser):
+        """
+        Initialize Scenario and validate inputs.
+
+        Parameter
+        ---------
+        parser
+            Parser object for reading in config files.
+
```
Comment (akey7) [inline celavi/scenario.py:42]: Consider putting a note about the exceptions this method raises, formatted like this:

```
Raises
-------
IOError
    Raises IOError when...
```

File: celavi/diagnostic_viz.py [hunk 2]
```diff
@@ -55,15 +56,19 @@ def __init__(
 
         value_name: str
             The name of the generalized value column, like 'count' or 'tonnes'.
+
+        run : int
+            Model run identifier for uncertainty runs within a scenario.
         """
         self.facility_inventories = facility_inventories
-        self.output_plot_filename = output_plot_filename
+        self.output_plot_filename = str.split(output_plot_filename, ".")[0] + '_' + str(run) + '.' + str.split(output_plot_filename, ".")[1]
```
Comment (akey7) [inline celavi/diagnostic_viz.py:64]: Consider something like the following for readability:

```
name, extension = str.split(output_plot_filename, ".")
self.output_plot_filename = f"{name}_{run}.{extension}"
```

File: celavi/costmethods.py [hunk 3]
```diff
@@ -259,10 +288,33 @@ def fine_grinding(path_dict):
...
+            # Parameters for fine grinding revenue
+            _c_fgr = path_dict['triang_params']['fine_grinding_revenue']['c']
+            _loc_fgr = path_dict['triang_params']['fine_grinding_revenue']['loc']
+            _scale_fgr = path_dict['triang_params']['fine_grinding_revenue']['scale']
+
+            _cost_unc = st.triang.rvs(
+                c=_c_fg, loc=_loc_fg * _cost, scale=_scale_fg * _cost
+            ) + st.triang.rvs(
+                c=_c_lf, loc=_loc_lf * _landfill, scale=_scale_lf * _landfill
+            ) + st.triang.rvs(
```
Comment (jwalzberg) [inline celavi/costmethods.py]: Should it be "minus" st.triang.rvs("revenue") and not "plus"?

**Review by akey7 — APPROVED** [review]
I haven't tried merging into develop yet. This branch runs on my machine. I left a few comments in the PR that you can implement or not as you see fit. Good job!

**Review by jwalzberg — APPROVED** [review]
@rjhanes I approved (code looks good regarding uncertainty), however, It doesn't work on my machine: 

- I followed the instruction regarding the branch of the celavi-data repo to use and created a "results" folder. 
- I used casestudy.yaml and scenario.yaml. 

I get an error after the year 2012 that seems to be related to the pylca_interface_process module. I pasted a screenshot of the error I get below:

image

**Comment by rjhanes** [comment]
@akey7  and @jwalzberg Thank you for the feedback! I created a second branch for this merge, `uncertaintypoc-merge`, that combines the fixes in issue #141 , the transportation disaggregation in #8 , and the uncertainty refactoring. 
Alicia - I implemented your suggested changes in that branch.
Julien - I think the error is due to the old format of the lca_db file; the number of columns changed with the addition of the FIPS and state columns. If you want to try deleting that file and re-running I think that will resolve the issue.

I'll be closing this PR shortly in favor of #152 which is undergoing final testing now.

**Comment by jwalzberg** [comment]
@rjhanes thanks for explaining the issue. The two runs went fine after deleting the lca_db file. It took about 20000 s for the first run and about 2000 s for the second.

**Pull request closed without merging**